### ISSUE TITLE
Enable payload change in the page_pre_render hook

### ISSUE THREAD

**Issue by jfayad** [Geeks-Solutions/nuxt-sections#97]
We shall have a way to let host projects read and alter the payload returned by the page API right before the page is rendered.

The effect of that change should only be visible in the UI and the changes should be not be persisted when saving the page

**Comment by akramsarkis** [comment]
How to test this @jfayad  ?  what changes will be done in the UI ?

**Comment by mahfouzhseineweev** [comment]
Automated tests:

* Write a test to make sure the views are the one the one coming from non persisted data version of the payload

**Comment by jfayad** [comment]
> How to test this @jfayad ? what changes will be done in the UI ?

Given you have one sections project with two domains
And each domain points to two separate front projects using the library, project A and Project B
And project A implements this hook to inject a section in the page while project B does NOT implement the hook
When you render the page on project A you see the additional section (eventhough it's not part of the API response)
And When you render the page on project B you see no additional section
And When you edit the page on project A, the additional section does NOT get stored in your page
And When you render the page on project B you still do NOT see the additional section (not in the page nor in the API response)

**Comment by akramsarkis** [comment]
@mahfouzhsein I cannot edit this issue but here the scope and the acceptance testing:

Given you have one sections project with two domains
And each domain points to two separate front projects using the library, project A and Project B
And project A implements this hook to inject a section in the page while project B does NOT implement the hook

- [ ] When you render the page on project A you see the additional section (eventhough it's not part of the API response)
- [ ] And When you render the page on project B you see no additional section
- [ ] And When you edit the page on project A, the additional section does NOT get stored in your page
- [ ] And When you render the page on project B you still do NOT see the additional section (not in the page nor in the API response)

### Acceptance testing

- [ ] When you render the page on project A you see the additional section (eventhough it's not part of the API response)
- [ ] And When you render the page on project B you see no additional section
- [ ] And When you edit the page on project A, the additional section does NOT get stored in your page
- [ ] And When you render the page on project B you still do NOT see the additional section (not in the page nor in the API

**Comment by mahfouzhseineweev** [comment]
We will be using a computed property for the altered views that will be have the return of a function that takes the initial views data property passed to the hook and returned to the page.

This way whenever the initial views data property changed (Edited in the forms) it will trigger the computed property to rerun and show the changes and altered view coming from the hook

**Comment by mahfouzhseineweev** [comment]
### Report:

The issue was supposed to be finished in a shorter amount of time but due to the complexity I faced in the computed properties, to have only the original content updated, this plus the fact that also the layout views management needed an update that I haven't considered initially

**Comment by mahfouzhseineweev** [comment]
@akramsarkis To test this issue I created a test hook file (which I should remove once you are done with testing):

The hook only works for this domain name `projectcnamenov.k8s-dev.geeks.solutions`

It includes two condition:

An update on the Home page `/` (A new section added)

And update on a page named `test1` (A new section added to the region of the layout)

**Comment by mahfouzhseineweev** [comment]
@akramsarkis I have the following questions

1. You tested on the `/` page that is not using a layout ? All good there ?

2. I checked for the first 2 points, it seems there are related only to the page creation. But then after the page is created and the layout is selected everything works well and you can sections to any region you want. You confirm ? (It is only when you directly edit the page after creating it without refreshing)

3. Can you please try the first 2 points on Maiori website ? Because I think both points are not introduced by the update I did but were already there

4. For the third point, I didn't get it at all and it seems your sentence is not completed

**Comment by akramsarkis** [comment]
> @akramsarkis I have the following questions
> 
> 1. You tested on the `/` page that is not using a layout ? All good there ?
> 2. I checked for the first 2 points, it seems there are related only to the page creation. But then after the page is created and the layout is selected everything works well and you can sections to any region you want. You confirm ? (It is only when you directly edit the page after creating it without refreshing)
> 3. Can you please try the first 2 points on Maiori website ? Because I think both points are not introduced by the update I did but were already there
> 4. For the third point, I didn't get it at all and it seems your sentence is not completed

@mahfouzhseineweev 

1. Yes I tested on / but / had already sections in it that I am using so i was not able to try while creating the page and adding sections
2. Yes I confirm, it is just when creating the page
3. I cannot as all the pages are used and I cannot create a new one
4. 4. I updated the sentence, in short the hooks section was showing only after saving and refreshing the page, for me it should show when I save the page without refreshing it

**Comment by akramsarkis** [comment]
all good now for this one, I do not have the right to close it

**Comment by akramsarkis** [comment]
@mahfouzhseineweev I will let the hook section for now, in case julien wants to see it before deploying
